### ISSUE TITLE
JsonCodable macro field names

### ISSUE THREAD

**Issue by hanskokx** [dart-lang/language#3829]
The [`JsonCodable`](https://dart.dev/language/macros#the-jsoncodable-macro) macro is a colossal step in the right direction, however it lacks an important feature which should be considered. Often, when working with serialized data (in particular, data returned from an API), the field names may differ from the appropriate Dart names. For instance, I've worked with APIs (particularly with Python backends) that would return JSON that looks like this:

```json
{
  "my_value": "some data"
}
```

When using Freezed or JsonSerializable, I could simply annotate the Dart object as thus:

```dart
@freezed
class AccessToken with _$AccessToken {
  const factory AccessToken({
    @JsonKey(name: 'access_token') required String accessToken,
    @JsonKey(name: 'refresh_token') required String refreshToken,
  }) = _AccessToken;

  factory AccessToken.fromJson(Map<String, dynamic> json) =>
      _$AccessTokenFromJson(json);
}
```

This allows me to map the snake case JSON value to my camel case Dart objects.

**Comment by jakemac53** [comment]
Yes, it lacks any configurability at all right now. This is because the metadata introspection APIs for macros are not yet implemented (and we are hashing out the details still).

Once it is well supported, we will certainly add this feature (as well as some other configuration options).
